File: problems/2024/x/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

float get_number_of_words_in_text(char text[]);

float get_avg_letters_in_text(string text, float words_count);

float get_avg_sentence_in_text(char text[], float words_count);

void calculate_readability(int score);

int main(void)
{
    string text = get_string("");
    float words_count = get_number_of_words_in_text(text);
    float avg_letters = get_avg_letters_in_text(text, words_count);
    float avg_sentence = get_avg_sentence_in_text(text, words_count);
    int index_score = (int) round(0.0588 * avg_letters - 0.296 * avg_sentence - 15.8);

    calculate_readability(index_score);

    return 0;
}

float get_number_of_words_in_text(char text[])
{
    float words_count = 1.0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isblank(text[i]))
        {
            words_count += 1;
        }
    }
    return words_count;
}

float get_avg_letters_in_text(string text, float words_count)
{
    float avg_letters = 0.0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (isalpha(text[i]))
        {
            avg_letters += 1.0;
        }
    }
    return (avg_letters / words_count) * 100;
}

float get_avg_sentence_in_text(char text[], float words_count)
{
    float avg_sentence = 0.0;
    for (int i = 0; i < strlen(text); i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            avg_sentence += 1.0;
        }
    }
    return (avg_sentence / words_count) * 100;
}

void calculate_readability(int score)
{
    if (score < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (score <= 2)
    {
        printf("Grade 2\n");
    }
    else if (score > 2 && score <= 16)
    {
        printf("Grade %d\n", score);
    }
    else
    {
        printf("Grade 16+\n");
    }
}

File: problems/2025/x/sentimental-readability/readability.py
import math


def get_number_of_words_in_text(text: str) -> float:
    return len(text.split())


def get_avg_letters_in_text(text: str, words_count: float) -> float:
    avg_letters = 0.0
    for letter in text:
        if letter.isalpha():
            avg_letters += 1.0

    return (avg_letters / words_count) * 100


def get_avg_sentence_in_text(text: str, words_count: float) -> float:
    avg_sentence = 0.0
    for sentence in text:
        if sentence == "." or sentence == "!" or sentence == "?":
            avg_sentence += 1.0

    return (avg_sentence / words_count) * 100


def calculate_readability(score: float):
    if score < 1:
        print("Before Grade 1")
    elif score <= 2:
        print("Grade 2")
    elif 2 < score <= 16:
        print(f"Grade {score}")
    else:
        print("Grade 16+")


def main():
    text = input("Text: ")
    words_count = get_number_of_words_in_text(text)
    avg_letters = get_avg_letters_in_text(text, words_count)
    avg_sentence = get_avg_sentence_in_text(text, words_count)

    index_score = int(round(0.0588 * avg_letters - 0.296 * avg_sentence - 15.8))
    calculate_readability(index_score)


if __name__ == "__main__":
    main()
